nixos/modules/config/users-groups.nix

```
@@ -419,12 +423,7 @@ let
           initialPassword initialHashedPassword;
         shell = utils.toShellPath u.shell;
       }) cfg.users;
-    groups = mapAttrsToList (n: g:
-      { inherit (g) name gid;
-        members = g.members ++ (mapAttrsToList (n: u: u.name) (
-          filterAttrs (n: u: elem g.name u.extraGroups) cfg.users
-        ));
-      }) cfg.groups;
+    groups = lib.attrNames cfg.groups;
```

I believe you meant `lib.attrValues`. Also since `with lib` is at the top of the file, there is no need to have `lib.attrValues`
```suggestion
    groups = attrValues cfg.groups;
```

---

@hqurve Since documentation is low, can you give me your definition of what the difference between attrNames and attrValues is, such that I can learn and others who read this PR can learn? Thanks!

---

Okay, sure.

Firstly, Ill explain what attrNames is. It is a builtin function which takes an attributeSet and returns a list containing the keys in alphabetical order. Example taken from the nix manual: 
```
builtins.attrNames { y = 1; x = "foo"; } == ["x" "y"]
```
On the other hand `attrValues` takes an attribute set and returns its values in sorted by the key they were associated with. This function could be defined as follows 
```
attrValues = attrs: map (key: attrs."${key}") (builtins.attrNames attrs)
``` 
And as a corresponding example
```
builtins.attrNames { y = 1; x = "foo"; } == ["foo" 1]
```
Note that these functions are also present in `lib` and `lib.attrsets` (I assume for convenience)
- lib
  - https://github.com/NixOS/nixpkgs/blob/f626a23cd3b512410f8d14a95b7cf94bba345900/lib/default.nix#L62 and https://github.com/NixOS/nixpkgs/blob/f626a23cd3b512410f8d14a95b7cf94bba345900/lib/default.nix#L75
- lib.attrsets
  - https://github.com/NixOS/nixpkgs/blob/f626a23cd3b512410f8d14a95b7cf94bba345900/lib/attrsets.nix#L12
  - https://github.com/NixOS/nixpkgs/blob/f626a23cd3b512410f8d14a95b7cf94bba345900/lib/attrsets.nix#L94

Now, onto why I suggested this change. Before this change, as you can see, `groups` contained a list of attribute sets, each containing attributes `name`, `gid` and `members`. Also, `cfg.groups` is an attribute set where each value contains `name`, `gid` and now also `members`. For example, say we have

```
cfg.groups = {
   wheel = {
      gid =1000;
      name = "wheel"; # This is defaulted to "wheel" if undefined (I dont understand the mechanics as to how)
      members = ["hqruve" "MattthewCroughan" "Infinisil"];
   };
   cool = {
      gid =1001;
      members = ["hqruve" "MatthewCroughan"];
   };
};
```

By leaving `groups = attrNames cfg.groups`, we would only get the group names. That is, `groups == ["cool" "wheel"]`. However, what we really want are the sets themselves. That is
```
groups == [
   {
      gid =1000;
      name = "wheel"; # This is defaulted to "wheel" if undefined (I dont understand the mechanics as to how)
      members = ["hqruve" "MattthewCroughan" "Infinisil"];
   }
   {
      gid =1001;
      members = ["hqruve" "MatthewCroughan"];
   }
]
```
and what function does this? `attrValues`

Also, you can see that r-rmcgibbo failed to build. Also note that there are multiple errors complaining that groups do not exist and `/etc/group` was not found. (I think the builder script for the groupfile either silently failed or threw an error Im unable to find)